File: benchmarks/micro/BranchSelectPredication.c
#include <stdint.h>
#include <unistd.h>

__attribute__((noinline)) static int64_t choose_value(int flag, int64_t left, int64_t right) {
    return flag ? left : right;
}

int main(void) {
    int64_t seed = (int64_t)getpid();
    int64_t total = seed & 1023;

    for (int32_t i = 0; i < 500000; i += 1) {
        int64_t wide = (int64_t)i;
        int flag = ((wide + total) & 1) == 0;
        total += choose_value(flag, wide + 3, total - wide);
        total %= 1000000007;
    }

    return total == 0 ? 1 : 0;
}

File: benchmarks/micro/BranchSelectPredication.rs
#[inline(never)]
fn choose_value(flag: bool, left: i64, right: i64) -> i64 {
    if flag {
        left
    } else {
        right
    }
}

fn main() {
    let seed = std::process::id() as i64;
    let mut total = seed & 1023;

    for i in 0_i32..500_000 {
        let wide = i as i64;
        let flag = ((wide + total) & 1) == 0;
        total += choose_value(flag, wide + 3, total - wide);
        total %= 1_000_000_007;
    }

    if total == 0 {
        std::process::exit(1);
    }
}
